### ISSUE TITLE
SHIFT + KEY and special chars don't work in text inputs

### ISSUE THREAD

**Issue by digitalLumberjack** [recalbox/recalbox-os#37]
Hard to input a ssid key without SHIFT in network settings

**Comment by hstampfl** [comment]
A small workaround: you can use CAPS LOCK to get the needed capital  letters

**Comment by digitalLumberjack** [comment]
Yep but it doesn't work for special characters...

**Comment by digitalLumberjack** [comment]
Someone have an idea ?

**Comment by hstampfl** [comment]
It's not only in the networksettings, I just tried it in the scraper. So, it seems to be globally.
Emulationstation has a keyboard input active with "Enter=Select" and also some buttons managed. Could it be that this mapping is not deactivated properly during the "Edit" mode?

**Comment by Churchill92** [comment]
This issue has to do with the keyboard/language mapping. Because the keyboard defaults to English(GB) instead of English(US) those functions (shift, underscore, dollar sign) do not seem to be present.  

 What needs to happen is to support ALL the keyboard/language packs from the upstream providers and run with it. Perhaps this needs to be coordinated with EmulationStation or not. I know raspi-config allows you to change all the keyboards/language without a problem.

**Comment by digitalLumberjack** [comment]
No it's not that,

If you switch to the terminal : F4 and ALT+F2 and connect with credentials "root" "recalboxroot" you will see that your shift key and will work. The default keyboard is US. 
You can use `loadkeys us` if you want to switch to us manually, but that will not change anything in emulationstation.

I have a fr keyboard, i load the keymap and it works everywhere except in emulationstation.

**Comment by hstampfl** [comment]
I am not really the programmer ;) but I would search the wrong staff in the SDL implementation...
In https://github.com/digitalLumberjack/recalbox-emulationstation/blob/recalbox-buildroot/es-core/src/InputManager.cpp from line 219 to 245. For me it looks like, that the SDL keyconversation  goes somewhere wrong. -> I do not have installed any c++ editor, so I can't go deeper in a short time

**Comment by digitalLumberjack** [comment]
What is wrong is that it works if i compile emulationstation (the recalbox
version) on a raspbian. So either it is something missing at
emulationstation compilation, either it is SDL2 that lack something.

I'll take a look, but i have not much time now. The workaround, for the
wifi at least, is in the next release.

**Comment by digitalLumberjack** [comment]
Cannot clode this one yet.
@nadenislamarre your skills are required :P

**Comment by nadenislamarre** [comment]
Looking on it tonight.

The conversion is done by X when you run es under X
by SDL
in SDL_x11events.c
#ifdef X_HAVE_UTF8_STRING
            if (data->ic) {
                X11_Xutf8LookupString(data->ic, &xevent.xkey, text, sizeof(text),
                                  &keysym, &status);
            }
#else
            XLookupString(&xevent.xkey, text, sizeof(text), &keysym, NULL);
#endif
            if (*text) {
                SDL_SendKeyboardText(text);
            }

When X is not present, it's far less simple :
it's in SDL_evdev.c
                                if (ioctl(_this->console_fd, KDGKBENT, (unsigned long)&kbe) == 0 && 
                                    ((KTYP(kbe.kb_value) == KT_LATIN) || (KTYP(kbe.kb_value) == KT_ASCII) || (KTYP(kbe.kb_value) == KT_LETTER))) 
                                {
				  kval = KVAL(kbe.kb_value, 0);
                                    if ( modstate & KMOD_CAPS && isalpha(kval) ) {
                                        if ( isupper(kval) ) {
					  kval = tolower(kval);
                                        } else {
					  kval = toupper(kval);
                                        }
                                    }
                                     
                                    /* Convert to UTF-8 and send */
                                    end = SDL_UCS4ToUTF8( kval, keysym);
                                    *end = '\0';

So, basically, i would say that something is missing under SDL on embeded device (or linux not having X) : SDL_INPUT_LINUXKD

I've not finished my investigation.
At least, i understand why i can't type in lowercase too ;-)
In the worst case, i think it's possible to make a basic patch to handle lower/upper and numbers.
Basically, a mapping conversion with the keyboard is missing. But i guess it exists while it works under bash for example.
I'll check how it works / or not / in kodi when the tv will be free.

**Comment by nadenislamarre** [comment]
So,
good news. It works in kodi15 on recalbox.
Kodi doesn't use sdl for keyboard (as far as i understand)

Kodi seems to get values using ioctl
  value = KeyboardGetSymbol(KeyboardReadValue(K_SHIFTTAB, entry.code));

unsigned short CLinuxInputDevice::KeyboardReadValue(unsigned char table, unsigned char index)
{
  struct kbentry entry;

  entry.kb_table = table;
  entry.kb_index = index;
  entry.kb_value = 0;

  if (ioctl(m_vt_fd, KDGKBENT, &entry)) ...
}

it's a bit late tonight, but i've a hope it will work soon.
I guess we have to do the same ioctl in sdl... but it's in fact more complicated
while ioctl are not on the keyboard, but on the tty

      m_vt_fd = open("/dev/tty0", O_RDWR | O_NOCTTY);

...

the kodi code even handles the keyboard leds ...
    SetLed(LED_CAPSL, 0);

so basically, all the code we need is in kodi.
The difficulty is to add it in sdl without any bad effect...

**Comment by nadenislamarre** [comment]
to resume :
when you're on a tty, you use the keymaps implemented by the tty
when you're on X, you use the keymaps implemented by x
otherwise, do as kodi does, call tty code conversion to switch keys depending on shift, alt.

i can continue on this issue.

**Comment by nadenislamarre** [comment]
in fact, it is what is exactly done by sdl.
but it fails.
it doesn't return the correct values.
So it seems that the keymaps are not the good one.

**Comment by nadenislamarre** [comment]
KARABOUMGA !
ca marche dans ma recalbox !

encore quelques points à vérifier (et à comprendre également), mais le patch ne devrait plus tarder.

**Comment by digitalLumberjack** [comment]
Great. Giga great

**Comment by digitalLumberjack** [comment]
i close this one :)

**Comment by wraybowling** [comment]
which release of 3.3.0 was this fixed for? I'm in beta17 and just bumped in to this problem.
